nemo/core/neural_interface.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+
+# =============================================================================
+# Copyright (c) 2020 NVIDIA. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+from abc import ABC, abstractmethod
+from typing import Dict, Optional
+
+import nemo
+from nemo.core.neural_types import NeuralType
+
+
+class NeuralInterface(ABC):
+    """
+        Abstract class offering interface shared between Neural Module and Neural Graph.
+        Had to move it to a separate class to:
+        a) avoid circular imports between Neural Module and Graph.
+        b) avoid collection of init_params implemented by default in Neural Module.
+        c) extract only the methods that are shared (NMs have plenty of methods that are not making any sense for
+        graph, e.g. get_weights, tie_weights, )
+    """
+
+    def __init__(self, name):
+        """ Constructor. Sets the application state. """
+        # Copy the name. As names should be unique in module/graph scope, this should be handled additionally
+        # in their constructors.
+        self._name = name
+        # Create access to the app state.
+        self._app_state = nemo.utils.app_state.AppState()
```

Comment:
Since AppState is a singleton, is it necessary to have to self variable refer to it? Isn't it always accessible at nemo.utils.app_state.AppState()?
Is this a style choice?

Reply:
this is only a "shortcut" that I personally find useful